src/Native/Runtime/unix/UnixEvent.cpp

```diff
@@ -0,0 +1,334 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+//
+// Implementation of events for Unix
+//
+
+#include "CommonTypes.h"
+#include "PalRedhawkCommon.h"
+#include "CommonMacros.h"
+#include "sal.h"
+#include "config.h"
+#include "PalRedhawkInline.h"
+#include "UnixCommon.h"
+#include <errno.h>
+#include <stdio.h>
+#include <string.h>
+#include <pthread.h>
+#include <sys/time.h>
+#ifdef __APPLE__
+#include <mach/mach.h>
+#include <mach/mach_time.h>
+#endif // __APPLE__
+#include "daccess.h"
+#include "slist.h"
+#include "slist.inl"
+
+#include "UnixEvent.h"
+
+static const uint32_t INFINITE = 0xFFFFFFFF;
+
+#define WAIT_OBJECT_0           0
+#define WAIT_TIMEOUT            258
+#define WAIT_FAILED             0xFFFFFFFF
+
+#ifdef __APPLE__
+// Convert nanoseconds to the timespec structure
+// Parameters:
+//  nanoseconds - time in nanoseconds to convert
+//  t           - the target timespec structure
+void NanosecondsToTimespec(uint64_t nanoseconds, timespec* t)
+{
+    t->tv_sec = nanoseconds / tccSecondsToNanoSeconds;
+    t->tv_nsec = nanoseconds % tccSecondsToNanoSeconds;
+}
+#endif // __APPLE__
+
+UnixEventWaiter::UnixEventWaiter(uint32_t eventCount)
+: m_eventIndex(-1),
+  m_timedOutOrFailed(false),
+  m_eventCount(eventCount)
+{
+    if (eventCount <= MaximumEmbeddedEvents)
+    {
+        m_waiterListEntries = m_embeddedWaiterListEntries;
+    }
+    else
+    {
+        FATAL_ASSERT(eventCount <= WAIT_TIMEOUT, "Too many events to wait on");
+
+        m_waiterListEntries = (ListEntry*)malloc(eventCount * sizeof(ListEntry));
```

Comment:
Unfortunatelly we need to be able to wait for more than one event, otherwise this change would not be needed at all, the existing code handled waiting for single event just fine (modulo the timed wait changes that would still be necessary). To be specific, we will need to wait for two events in the RhpWaitForFinalizerRequest once we add the low memory event for Unix. I've discussed this with @Maoni0 and she said it will be needed.
As for the dynamic allocation, it was a last minute change and I was going back and forth whether to make it or have the limit set to 2 events. In the end, I felt like it doesn't bring in much additional code or complexity and so I have ended up adding it. But I am fine removing it. We can easily add it any time if we find it needed.
Regarding the RhCompatibleReentrantWaitAny, I will add it to this change.